### ISSUE TITLE
RFE: Add a macro CGAL_UNREACHABLE in <CGAL/config.h>

### ISSUE THREAD

**Issue by lrineau** [CGAL/cgal#2921]
## Issue Details

(See the context at https://github.com/CGAL/cgal/pull/2912#issuecomment-371805213 and later.)

Add a macro `CGAL_UNREACHABLE` that could be used for example as:
```c++
switch(orientation) {
   //case ...
  default: CGAL_UNREACHABLE;
};
```
See also the C++ proposal [PR0627](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2017/p0627r1.pdf).

## Source Code

Not tested.
```C++
// Macro CGAL_ASSUME
// Call a builtin of the compiler to pass a hint to the compiler
#if __has_builtin(__builtin_unreachable) || (CGAL_GCC_VERSION >= 40500 && !__STRICT_ANSI__)
// From g++ 4.5, there exists a __builtin_unreachable()
// Also in LLVM/clang
#  define CGAL_ASSUME(EX) if(!(EX)) { __builtin_unreachable(); }
#  define CGAL_UNREACHABLE __builtin_unreachable()
#elif defined(_MSC_VER)
// MSVC has __assume
#  define CGAL_ASSUME(EX) __assume(EX)
#  define CGAL_UNREACHABLE __assume(false)
#else
// If CGAL_ASSUME is not defined, then CGAL_assume and CGAL_assume_code are
// defined differently, in <CGAL/assertions.h>
#  define CGAL_UNREACHABLE CGAL_error_msg("unreachable branch")  
#endif
```

**Comment by lrineau** [comment]
Actually, we also need a macro named, for example `CGAL_DEFAULT_CASE_IS_UNREACHABLE`, because otherwise the code
```c++
default: CGAL_UNREACHABLE;
```
is not optimal when the compiler does not have `__builtin_unreachable`/`__assume`. With those compilers, it is best not having a `default` case at all (so `CGAL_DEFAULT_CASE_IS_UNREACHABLE` would be empty, in that case).

**Comment by lrineau** [comment]
> The macros have been created by commit [85c0035](https://github.com/CGAL/cgal/commit/85c00352c39e12a9d094de50aead311ceb6dccac)

That was in #6052 (merged in 5.4).

As for the comment https://github.com/CGAL/cgal/issues/2921#issuecomment-372388364:
> Actually, we also need a macro named, for example `CGAL_DEFAULT_CASE_IS_UNREACHABLE`, because otherwise the code
> 
> ```c++
> default: CGAL_UNREACHABLE;
> ```
> 
> is not optimal when the compiler does not have `__builtin_unreachable`/`__assume`. With those compilers, it is best not having a `default` case at all (so `CGAL_DEFAULT_CASE_IS_UNREACHABLE` would be empty, in that case).

All compilers we support have either `__builtin_unreachable` (clang/gcc) or `__assume` (MSVC). Anyway, [`[[assume]]`](https://en.cppreference.com/w/cpp/language/attributes/assume) and [`std::unreachable`](https://en.cppreference.com/w/cpp/utility/unreachable) were standardized in C++23. That means that eventually all compilers will support them. The doc page of `std::unreachable` even give as an example the `default:` label of a switch:
```c++
    default:
        std::unreachable();
```
